cpp/src/strings/substring.cu

```diff
@@ -257,30 +249,111 @@ struct dispatch_substring_from_fn {
-}
+// When slice_strings is invoked with a delimiter string and a delimiter count, we need to
+// compute the start and end indices of the substring. This functor accomplishes that.
+struct compute_substring_indices {
+  template <typename DelimiterItrT>
+  void operator()(column_device_view const& d_column,
+                  DelimiterItrT const delim_itr,
+                  size_type delimiter_count,
+                  size_type* start_char_pos,
+                  size_type* end_char_pos,
+                  rmm::mr::device_memory_resource* mr,
+                  cudaStream_t stream) const
+  {
+    auto strings_count = d_column.size();
+
+    thrust::for_each_n(
+      rmm::exec_policy(stream)->on(stream),
+      thrust::make_counting_iterator<size_type>(0),
+      strings_count,
+      [delim_itr, delimiter_count, start_char_pos, end_char_pos, d_column] __device__(
+        size_type idx) {
+        auto const& delim_val_pair = delim_itr[idx];
+        auto const& delim_val      = delim_val_pair.first;  // Don't use it yet
 
+        // If the column value for this row is null, result is null.
+        // If the delimiter count is 0, result is empty string.
+        // If the global delimiter or the row specific delimiter is invalid or if it is empty, row
+        // value is empty.
+        if (!d_column.is_null(idx) && delim_val_pair.second && !delim_val.empty()) {
+          auto const& col_val = d_column.element<string_view>(idx);
+
+          // If the column value for the row is empty, the row value is empty.
+          if (!col_val.empty()) {
+            auto const col_val_len   = col_val.length();
+            auto const delimiter_len = delim_val.length();
+
+            auto nsearches      = (delimiter_count < 0) ? -delimiter_count : delimiter_count;
+            size_type start_pos = 0;
+            size_type end_pos   = col_val_len;
+            bool keep_searching = true;
+
+            for (auto i = 0; keep_searching && i < nsearches; ++i) {
+              if (delimiter_count < 0) {
+                end_pos = col_val.rfind(delim_val, 0, end_pos);
+                if (end_pos == -1) {
+                  start_char_pos[idx] = 0;
+                  end_char_pos[idx]   = col_val_len;
+                  keep_searching      = false;
+                } else if (i + 1 == nsearches) {
+                  start_char_pos[idx] = end_pos + delimiter_len;
+                  end_char_pos[idx]   = col_val_len;
+                }
+              } else {
+                auto char_pos = col_val.find(delim_val, start_pos);
+                if (char_pos == -1) {
+                  start_char_pos[idx] = 0;
+                  end_char_pos[idx]   = col_val_len;
+                  keep_searching      = false;
+                } else if (i + 1 == nsearches) {
+                  start_char_pos[idx] = 0;
+                  end_char_pos[idx]   = char_pos;
+                } else
+                  start_pos = char_pos + delimiter_len;
+              }
```

Comment:
There's a lot of repetition in this loop. I rewrote it to remove duplication, it _should_ do the same:
```suggestion
start_char_pos[idx] = 0;
end_char_pos[idx] = col_val_len;
size_type char_pos = -1;
for (auto i = 0; i < nsearches; ++i) {
  char_pos = left_to_right ? col_val.find(delim_val, start_pos)
                 : col_val.rfind(delim_val, 0, end_pos);
  if (char_pos == -1)
    return;
  if (left_to_right)
    start_pos = char_pos + delimiter_len;
  else
    end_pos = char_pos;
}
if (left_to_right)
  end_char_pos[idx] = char_pos;
else
  start_char_pos[idx] = end_pos + delimiter_len;
```

Reply:
done